In [1]:
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn import preprocessing
from sklearn.ensemble import RandomForestClassifier

# 1. Titanic Dataset
- Build Decision Tree(DV-"Survived",IDV-"Age,Gender and Fare") and Prediction

### Clasification

In [2]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')

In [3]:
df_train['Age'].mean()

32.69985376827896

In [4]:
train_age = np.where(df_train['Age'].isnull(),32,df_train['Age'])

In [5]:
df_test['Age'].mean()

33.23741007194245

In [6]:
test_age = np.where(df_test['Age'].isnull(),33,df_test['Age'])

In [7]:
label_en = preprocessing.LabelEncoder()

In [8]:
train_gender = label_en.fit_transform(df_train['Sex'])

In [9]:
test_gender = label_en.fit_transform(df_test['Sex'])

In [10]:
Features = pd.DataFrame([train_age,df_train['Fare'],train_gender]).T

In [11]:
tree_model = tree.DecisionTreeClassifier(max_depth=6)

In [12]:
tree_model.fit(X = Features, y= df_train['Survived'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=6, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [13]:
tree_model.score(X=Features, y = df_train['Survived'])

0.8256467941507312

> we get model score as 82.56 % when the Features are Age, Fare and Gender

### Predication

In [17]:
Features_test = pd.DataFrame([test_age,df_test['Fare'],test_gender]).T

In [19]:
test_predict = tree_model.predict(X=Features_test)

In [21]:
predicated_output = pd.DataFrame({"P_Id": df_test['PassengerId'],"Survived": test_predict})

In [22]:
predicated_output.head()

P_Id  Survived
0   892         0
1   893         1
2   894         0
3   895         0
4   896         1

In [24]:
with open('dtree1.dot','w') as f:
    f = tree.export_graphviz(tree_model, feature_names=['AGE','FARE','GENDER'], out_file= f)

# 2. Build Decision Tree for Attrition Rate Analysis

- DV - "Attrition"
- IDV - Output of RF Algorithm


In [2]:
df2 = pd.read_csv('Attrition_anaysis.csv')
df2.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [3]:
new_NumCompaniesWorked  = np.where(df2['NumCompaniesWorked'].isnull(), 2.69, df2['NumCompaniesWorked'] )
df2['NumCompaniesWorked'] = new_NumCompaniesWorked

In [4]:
new_TotalWorkingYears  = np.where(df2['TotalWorkingYears'].isnull(), 11.27, df2['TotalWorkingYears'] )
df2['TotalWorkingYears'] = new_TotalWorkingYears

In [5]:
le = preprocessing.LabelEncoder()

In [6]:
df3 = df2.copy()

In [7]:
df3['Attrition'] = le.fit_transform(df2['Attrition'])
df3['BusinessTravel'] = le.fit_transform(df2['BusinessTravel'])
df3['Department'] = le.fit_transform(df2['Department'])
df3['EducationField'] = le.fit_transform(df2['EducationField'])
df3['Gender'] = le.fit_transform(df2['Gender'])
df3['JobRole'] = le.fit_transform(df2['JobRole'])
df3['MaritalStatus'] = le.fit_transform(df2['MaritalStatus'])
df3['Over18'] = le.fit_transform(df2['Over18'])

In [8]:
rf_model = RandomForestClassifier(n_estimators=2000, max_features=2 , oob_score= True)

In [9]:
features_att =['BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Age',
         'DistanceFromHome', 'JobLevel', 'MonthlyIncome','NumCompaniesWorked', 'PercentSalaryHike',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany', 'YearsSinceLastPromotion', 
         'YearsWithCurrManager']

In [10]:
df3[features_att]


BusinessTravel  Department  EducationField  Gender  JobRole  \
0                  2           2               1       0        0   
1                  1           1               1       0        6   
2                  1           1               4       1        7   
3                  0           1               1       1        1   
4                  2           1               3       1        7   
...              ...         ...             ...     ...      ...   
4405               2           1               3       0        6   
4406               2           1               3       1        2   
4407               2           1               1       1        7   
4408               2           2               3       1        2   
4409               2           1               3       1        2   

      MaritalStatus  Age  DistanceFromHome  JobLevel  MonthlyIncome  \
0                 1   51                 6         1         131160   
1                 2   31                10         1          41890   
2                 1   32                17         4         193280   
3                 1   38                 2         3          83210   
4                 2   32                10         1          23420   
...             ...  ...               ...       ...            ...   
4405              2   42                 5         1          60290   
4406              0   29                 2         1          26790   
4407              1   25                25         2          37020   
4408              0   42                18         1          23980   
4409              0   40                28         2          54680   

      NumCompaniesWorked  PercentSalaryHike  StockOptionLevel  \
0                    1.0                 11                 0   
1                    0.0                 23                 1   
2                    1.0                 15                 3   
3                    3.0                 11                 3   
4                    4.0                 12                 2   
...                  ...                ...               ...   
4405                 3.0                 17                 1   
4406                 2.0                 15                 0   
4407                 0.0                 20                 0   
4408                 0.0                 14                 1   
4409                 0.0                 12                 0   

      TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                  1.00                      6               1   
1                  6.00                      3               5   
2                  5.00                      2               5   
3                 13.00                      5               8   
4                  9.00                      2               6   
...                 ...                    ...             ...   
4405              10.00                      5               3   
4406              10.00                      2               3   
4407               5.00                      4               4   
4408              10.00                      2               9   
4409              11.27                      6              21   

      YearsSinceLastPromotion  YearsWithCurrManager  
0                           0                     0  
1                           1                     4  
2                           0                     3  
3                           7                     5  
4                           0                     4  
...                       ...                   ...  
4405                        0                     2  
4406                        0                     2  
4407                        1                     2  
4408                        7                     8  
4409                        3                     9  

[4410 rows x 18 columns]

In [11]:
rf_model.fit(X=df3[features_att], y=df3['Attrition'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=2000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [12]:
print(rf_model.oob_score_)

1.0


In [13]:
for features_att, imp in zip(features_att, rf_model.feature_importances_):
    print(features_att,imp)

BusinessTravel 0.029069337295804923
Department 0.02723156680963429
EducationField 0.043066599516669704
Gender 0.01937112695302434
JobRole 0.05918593044555502
MaritalStatus 0.04086032749936947
Age 0.10034634128343199
DistanceFromHome 0.0737052529496861
JobLevel 0.03977788764882078
MonthlyIncome 0.09913006459773976
NumCompaniesWorked 0.0582525042498868
PercentSalaryHike 0.0686337154725921
StockOptionLevel 0.03527167560352816
TotalWorkingYears 0.08869296321712786
TrainingTimesLastYear 0.04766522315369618
YearsAtCompany 0.06998811824316722
YearsSinceLastPromotion 0.04477144078738997
YearsWithCurrManager 0.05497992427287529


- looking at the importance score for each feature we can conclude that Features like 
> age, Monthlyincome, Totalworkingyears , distancefromhome, percentagesalaryhike, yearsatcompany seems to be most imporatant among other features 

In [14]:
Features_test_attr = [ 'Age','DistanceFromHome',  'MonthlyIncome', 'PercentSalaryHike',
        'TotalWorkingYears', 'YearsAtCompany']

In [16]:
tree_model = tree.DecisionTreeClassifier(max_depth=12)

In [17]:
tree_model.fit(X = df3[Features_test_attr], y= df3['Attrition'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [18]:
tree_model.score(X=df3[Features_test_attr], y = df3['Attrition'])

0.9480725623582766

> With model score of 94.8% we can say that 'Age', 'DistanceFromHome', 'MonthlyIncome', 'PercentSalaryHike',
        'TotalWorkingYears', 'YearsAtCompany' features play important role in attrition level in the company

# 3. Build Decision Tree for Bank Loan Modelling

- DV - "Personal Loan"
- IDV - Output of RF Algorithm


In [21]:
dataset = pd.read_excel('Bank_Personal_Loan_Modelling.xlsx', sheet_name=1)
dataset.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [22]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


- no null values in the dataset 
- all features are in number form 
- columns lie ID, ZIP code can be negleted from analysis

In [30]:
rf_model2 = RandomForestClassifier(n_estimators=2000, max_features=2 , oob_score= True)

In [23]:
dataset.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [27]:
features_bank =[ 'Age', 'Experience', 'Income', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']

In [28]:
dataset[features_bank].head()


Age  Experience  Income  Family  CCAvg  Education  Mortgage  \
0   25           1      49       4    1.6          1         0   
1   45          19      34       3    1.5          1         0   
2   39          15      11       1    1.0          1         0   
3   35           9     100       1    2.7          2         0   
4   35           8      45       4    1.0          2         0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [31]:
rf_model2.fit(X=dataset[features_bank], y=dataset['Personal Loan'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=2000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [32]:
print(rf_model2.oob_score_)

0.9876


In [33]:
for features_bank, imp in zip(features_bank, rf_model2.feature_importances_):
    print(features_bank,imp)

Age 0.04476425924581377
Experience 0.04397100577426114
Income 0.3467924995731448
Family 0.09698293448232892
CCAvg 0.18170039513327313
Education 0.1649968072571762
Mortgage 0.04345668360005258
Securities Account 0.005592062015944805
CD Account 0.053284061052486564
Online 0.00852515758563117
CreditCard 0.009934134279886982


> we can obsevrve that  income, ccAvg, education, family, CD Account seem to be important features 

In [34]:
dataset.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

### model 1

In [35]:
Features_test_bank = [  'Family', 'CCAvg',
       'Education', 'CD Account']

In [37]:
tree_model = tree.DecisionTreeClassifier(max_depth=8)

In [38]:
tree_model.fit(X = dataset[Features_test_bank], y= dataset['Personal Loan'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=8, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [39]:
tree_model.score(X = dataset[Features_test_bank], y= dataset['Personal Loan'])

0.961

### model 2

In [60]:
Features_test_bank2 = [  'Family', 'CCAvg',
       'Education', 'CD Account','Age','Mortgage']

In [61]:
tree_model = tree.DecisionTreeClassifier(max_depth=12)

In [62]:
tree_model.fit(X = dataset[Features_test_bank2], y= dataset['Personal Loan'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [63]:
tree_model.score(X = dataset[Features_test_bank2], y= dataset['Personal Loan'])

0.9866